utilities/parameters.txt
# $Id: parameters.txt 20 2005-06-22 15:20:16Z brc $
# $Log$
# Revision 1.1  2005/06/22 15:19:54  brc
# Added to complete the contents of the repository.
#
# Revision 1.1.1.1  2003/02/03 20:18:44  brc
# This is the re-organized distribution of libccom (a.k.a. CUBE),
# which has a more realistic structure for future development.  The
# code re-organization and build system was contributed by IVS
# (www.ivs.unb.ca).
#
# Revision 1.1.2.3  2003/01/31 22:27:51  dneville
# (BRC) Modified eeg_q_limit in both locations to essentially turn off this piece
# of the code.  It isn't required any more, and needs good ad hoc calibration
# to work effectively in this mode.
#
# Revision 1.1.2.2  2003/01/28 16:38:45  dneville
# (BRC) Added 'sonar_type' element to global section of the vessel parameter
# clause, and a corresponding element in the Vessel description that
# allows it to be retained.  This allows one file, with just a Vessel clause, to
# specify the whole system being used on the survey platform (makes it easier
# for integrators to keep things together).
#
# Revision 1.1.2.1  2003/01/28 16:34:23  dneville
# Adding a new file that CUBE uses to store the default parameters for platform setup  (offsets, error model params, etc.. ).
#
# Revision 1.7.2.1  2002/12/15 01:58:29  brc
# Added defaults for the blunder filters.
#
# Revision 1.7  2002/05/10 22:18:27  brc
# Added parameter for probability that a node is going to be used under the
# HyperCUBE scheme.  This is used to adjust the size calculations.
#
# Revision 1.6  2002/03/14 04:44:48  brc
# Added interpolation module, CARIS/HIPS(HDCS) data formats and dual-head
# vessel specifications.  Removed static roll offsets from error model (now
# specified through the vessel description on a per-head basis as it always
# should have been ...)
#
# Revision 1.5  2001/12/07 21:26:05  brc
# Added 'static roll' element in error model code, and 'select' statement in
# GSF file reader.  The former allows for head roll offsets, typically for
# dual head arrangements, and the latter allows for selection of which head
# to report when reading dual-head data from GSF.
#
# Revision 1.4  2001/10/15 23:11:17  brc
# Added SoG default values in Simrad raw module section.
#
# Revision 1.3  2001/09/23 19:25:31  brc
# Added new parameters in nav and tides modules, added HyperCUBE and the
# separate set of parameters for the standalone version of CUBE.
#
# Revision 1.2  2001/05/14 20:45:45  brc
# Added mapsheet module parameters
#
# Revision 1.1  2001/05/14 04:15:22  brc
# Information and test parameter file for use with test_params.c
#
#
# Prototype parameter defaults file for libccom V1.0.
# Date: 13 May 2001.
#
# This file is used to test the parameter file reader module, and to provide
# full details of all of the parameters that are known in the various modules
# of the library.  Each module is identified by its top-level name, and sub-
# modules nested within them as appropriate.  The parameters are read into a
# uniform list, and are then passed to the top of the module hierarchy (what
# happens afterwards depends very much on the module; the top-level executive
# only sends it to the top of each tree).
#
# Notes:
#  1. Where there are a number of ways to describing a number, or where there
#     is the possibility of a scale factor, a unit may be given.  Lengths may
#     have 'nm' for nautical miles, 'm' for meters, 'ft' for feet, and 'in' for
#     inches; for MKS units, multipliers are allowed from 'u' to 'k'. Durations
#     have to be in seconds, but multipliers from 'u' to 'c' are allowed.  Rates
#     have to be in Hertz, multipliers 'k' and 'M' are allowed.
#  2. Angles are always given in degrees, so no units are required.  Ditto
#     speeds and m/s.
#  3. Some numbers have no units, and any units will probably be quietly ignored
#     by the reader.
#  4. The data tags have fairly free specification limits.  Numbers can be up
#     to full floats, and are always interpreted as floats.  Free form text is
#     allowed, including /, - and . so that filenames can (mostly) be specified.
#     The only limitation is that spaces are not allowed.
#  5. Each clause (module) is specified by an identifier, then a brace enclosed
#     list of sub-clauses and element lines.  Element lines are an identifier,
#     some data tag, and then a semi-colon.  The parser and lexer are white-
#     space tolerant, with the exception of comments '# <text>' which extend to
#     the end of the current line.
#  6. The parameter file reader doesn't really understand anything about the
#     module names, parameter names or allowed data.  It just reads the file
#     and then offers it back to the user; any error checking has to be done in
#     the individual modules.  The general executive in params_execute() knows
#     about the modules in the internal table in params.c, and when it is passed
#     a parameters list, it gives it to each module in turn and in table order.
#  7. The values in this file are intended to be the defaults that are hard
#     coded into the modules, so that the user can see what needs to be changed.
#     The modules use hard defaults if the parameters are not modified in the
#     parameter list, so only the values that need to be changed need to be in
#     this file.

# Generic interpolation module
interp {
	buffer_quantum 1024;
}

# Attitude interpolation module
att {
	buffer_quantum 1024;
}

# Navigational interpolation module
nav {
	buffer_quantum 1024;
	latency		   0 s;	# Set navigation latency.  Positive values indicate
						# that the navigation lags the data.
}

# Survey Time interpolation and translation module
stime {
	buffer_quantum 1024;
}

# Tide Correction interpolation and translation module
tides {
	buffer_quantum 256;
	time_offset		0 s;	# Time offset to apply on read: +ve => worksite lags
	scale_factor	1.0;	# Scale factor to apply on read (0.0,1.0].
}

# Error Model module
errmod {
	# Full MBES Error model
	full {
		default_surf_sspeed 1500.0;
	}
	# Simplified model based on IHO S44 4ed
	iho {
		order 		1;			# Set IHO survey order 1..4
		vert_fixed	20 cm;		# These specifically set the elements in the
		vert_var	0.0075;		# survey order table, but are unnecessary if the
		horizontal	2 m;		# survey matches the IHO table.  These allow
								# the entries to be over-ridden if required.
								#   The code uses 'order' to set base values,
								# and then modifies any which are specified.
	}
}

# Vessel configuration and auxilliary instrument measurement errors
vessel {
	timestamp	1970/01/01:00:00:00;
	sonar_type	unknown;	# device module recognition string for data generator
	draft		0 m;	# Transducer draft below RP
	tx_latency	0 ms;	# Transducer transmit latency
	static_roll	0;		# Head mounting angle, deg., +ve => port up
	gps {
		x		0 m;	# GPS 3D offsets from transducer
		y		0 m;
		z		0 m;
		latency	0 ms;	# GPS latency
	}
	imu {
		x		0 m;	# IMU 3D offsets from transducer
		y		0 m;
		z		0 m;
		roll	0;		# Alignment angles for IMU to transducer.
						# Note that angles are always given in degrees.
		pitch	0;
		gyro	0;
	}
	port {					# Port head definitions (optional)
		draft		0 m;	# Transducer draft below RP
		tx_latency	0 ms;	# Transducer transmit latency
		static_roll	0;		# Head mounting angle, deg., +ve => port up
		gps {
			x		0 m;	# GPS 3D offsets from transducer
			y		0 m;
			z		0 m;
			latency	0 ms;	# GPS latency
		}
		imu {
			x		0 m;	# IMU 3D offsets from transducer
			y		0 m;
			z		0 m;
			roll	0;		# Alignment angles for IMU to transducer.
							# Note that angles are always given in degrees.
			pitch	0;
			gyro	0;
		}
	}
	stbd {					# Starboard head definitions (optional)
		draft		0 m;	# Transducer draft below RP
		tx_latency	0 ms;	# Transducer transmit latency
		static_roll	0;		# Head mounting angle, deg., +ve => port up
		gps {
			x		0 m;	# GPS 3D offsets from transducer
			y		0 m;
			z		0 m;
			latency	0 ms;	# GPS latency
		}
		imu {
			x		0 m;	# IMU 3D offsets from transducer
			y		0 m;
			z		0 m;
			roll	0;		# Alignment angles for IMU to transducer.
							# Note that angles are always given in degrees.
			pitch	0;
			gyro	0;
		}
	}
	sdev {						# These are measurement errors for the various
								# auxilliary instruments in the survey platform.
		gps_offset		5 mm;	# All 3D GPS offsets
		gps_drms		2 m;	# Base GPS 2D horizontal uncertainty
		gps_latency		30 ms;
		imu_offset		5 mm;	# All 3D IMU offsets
		imu_rp_align	0.05;	# IMU Roll/Pitch alignment
		imu_g_align		0.05;	# IMU Gyro alignment
		imu_latency		5 ms;
		roll			0.05;	# Roll measurement error
		pitch			0.05;	# Pitch measurement error
		pitch_stab		0;		# Pitch stabilisation measurement error
								# i.e., here, no pitch stabilisation
		gyro			0.50;	# Gyro measurement error
		svp				0.52;	# Sound velocity profile measurement error.
								# Note that speeds are always given in m/s, so
								# no units are required
		surf_sspeed		0.50;	# Surface sound speed measurement error
		heave_fixed		5 cm;	# Fixed component of heave measurement error
		heave_var		0.05;	# Variable component of heave measurement error
		draft			2 cm;
		ddraft			2 cm;	# Dynamic draft (e.g., squat)
		loading			1 cm;	# Known loading (e.g., fuel use)
		sog				0.2;	# Speed over ground
		tide_meas		2 cm;	# Tide measurement error at gauge
		tide_pred		2 cm;	# Tide error due to spatial/temporal variation
	}
}

# Sounding provision/translation module
sounding {
	default_surf_sspeed		1500.0;		# Default surface sound speed, m/s
	gsf {
		default_surf_sspeed	1500;		# Same for GSF reader module
		select				any;		# Can be port, stbd or any to select
										# sonar head from GSF file.
	}
	hips {
		default_surf_sspeed	1500;		# Default surface sound speed, m/s
		amp_phase_trans		12;			# Number of samples in footprint needed
										# to jump to phase detection
	}
	omg {
		honour_flags		yes;		# yes|no - whether to use edit flags
		default_mean_speed	1500.0;		# Mean water column speed, m/s
		default_surf_sspeed	1500.0;		# Surface sound speed, m/s
		default_vessel_speed   3.0;		# Vessel speed, m/s
	}
	raw {
		buffer_exageration		3.0;	# Buffer safety factor (1.0, 10.0)
		simrad {
			max_resynch			1000;		# Maximum number of synch failures
											# before giving up
			att_op_rate			10.0 Hz;	# Output rate for decimated high-
											# rate attitude data
			att_failure_rate	1.0 Hz;		# Input rate of high-rate attitude
											# before deciding latencies reported
											# in file are erroneous
			att_min_rate		10.0 Hz;	# Minimum attitude input rate before
											# switching to pass-through FIR
			att_fir_len			51;			# Length of decimation FIR
			att_rate_slop		0.05;		# Fractional change in measured
											# input rate before changing the FIR
			default_sog			4.116;		# Ship's speed-over-ground if not
											# available from position datagram
											# (in m/s, range (0.0,15.4); 4.116
											# m/s is 8 kts, 15.4 m/s is 30 kts).
		}
	}
}

# MapSheet depth estimation and organisation module
mapsheet {
	max_height		10000;			# Maximum sizes in nodes
	max_width		10000;
	
	# Iterative Trimmed Mean Estimation Algorithm
	mean {
		null_depth			0 m;	# Reconstruction depth with no data
		reject_sdev_limit	2.0;	# Std. Dev. limit at which to cull outliers
		mean_ratio_eps		0.01;	# Mean ratio convergence limit
		sdev_ratio_eps		0.01;	# SDev ratio convergence limit
		max_iterations		100;	# Max. Iter. before giving up on convergence
		buffer_quantum		256;	# Buffer allocation quantum
		max_buffer_pt		65536;	# Maximum buffer size
	}
	
	# Median Bin Estimation Algorithm
	median {
		null_depth			0 m;	# Reconstruction depth with no data
		subsample_at		500;	# Median filtering sub-sample limit
		buffer_quantum		256;	# Buffer allocation quantum
		max_buffer_pt		65536;	# Maximum buffer size
	}
	
	# Concurrent Uncertainty and Bathymetry Estimation (CUBE) Algorithm
	nodal {
		null_depth			0 m;	# Reconstruction depth with no data
		null_sdev			1 km;	# Reconstruction sdev with no data
		confidence_int		C95;	# Confidence interval to report
									# Values: C68 (1 s.d.), C90, C95, C99
		distance_exp		2.0;	# Uncertainty dilution function exponent
		median_length		11;		# Median pre-queue length
		eeg_q_limit			255.0;	# Continuous Eeg-Clipping Quotient Limit
		evolution_discount	1.0;	# West & Harrison evolution variance
									# discount factor (range [0.8,1.0]).
		monitor_offset		4.0;	# Model monitoring vertical offset
									# (range (0.0, \infty))
		monitor_tau			0.135;	# Model monitoring Bayes factor limit
									# (range (0.0, \infty))
		monitor_runlength	5;		# Model monitoring `bad run' limit before
									# intervention.  Integer, positive
		extractor			likelihood;	# Hypothesis selection algorithm
										# Values: 'prior', 'likelihood' or
										# 'posterior'.
		min_context			5 m;	# Minimum search radius for guide node in
									# spatial context hypothesis selection
									# (i.e., in 'likelihood' and 'posterior'
									# selection modes).
		max_context			10 m;	# Maximum search radius for guide node
	}
	
	# The HyperCUBE algorithm: multiple CUBE sheets with LRU caching
	hcube {
		max_memory			128 Mb;	# Maximum amount of memory to use
		hypothesis_hint		1.25;	# Scale for Expected num. hypotheses
		prob_use			0.80;	# Probability a node is used (expected)
		max_tile_dim		511;	# Maximum size of a tile
		tile_expiry_delta	600 s;	# Time-out before flushing tiles
	}
}

# The base CUBE algorithm
cube {
		null_depth			0 m;	# Reconstruction depth with no data
		null_sdev			1 km;	# Reconstruction unct. with no data
		confidence_int		C95;	# Report 95% confidence intervals
		distance_exp		2.0;	# User exponent for dilution function
		median_length		11;		# Size of median pre-queue in node
		eeg_q_limit			255.0;	# Continual culling with Egg's Quotients
		evolution_discount	1.0;	# West & Harrison evolution variance
									# discount factor (range [0.8,1.0])
									# 1.0 = no discount (max. smoothing).
		monitor_offset		4.0;	# Model monitoring vertical offset limit
									# (range (0.0, \infty))
		monitor_tau			0.135;	# Model monitoring Bayes factor limit for
									# single and sequential monitoring.
									# (range (0.0, \infty)).
		monitor_runlength	5;		# Model monitoring `bad run' limit before
									# intervention; integer, positive.
		extractor			likelihood;	# Disambiguation method; choices are
										# 'prior', 'likelihood', 'posterior'
		min_context			5 m;	# Minimum disambiguation search limit
		max_context			10 m;	# Maximum disambiguation search limit
		blunder_min			10 m;	# Minimum blunder threshold (m)
		blunder_pcent			25.0;	# Percentage of depth to consider a blunder
							# (in deep direction only).
		blunder_scalar			3.0;	# Scale of initialisation surface variance to allow
							# before declaring deeps to be blunders
}


libsrc/ccom_core/params.c
/*
 * $Id: params.c 2 2003-02-03 20:18:41Z brc $
 * $Log$
 * Revision 1.1  2003/02/03 20:18:44  brc
 * Initial revision
 *
 * Revision 1.2.4.1  2003/01/28 14:29:27  dneville
 * Latest updates from Brian C.
 *
 * Revision 1.7.2.1  2002/07/14 02:20:47  brc
 * This is first check-in of the Win32 branch of the library, so that
 * we have a version in the repository.  The code appears to work
 * on the Win32 platform (tested extensively with almost a billion data
 * points during the GoM2002 cruise on the R/V Moana Wave,
 * including multiple devices operating simultaneously), but as yet,
 * a Unix compile of the branch has not been done.  Once this is
 * checked, a merge of this branch to HEAD will be done and will
 * result in a suitably combined code-base.
 *
 * Revision 1.7  2002/03/14 04:04:22  brc
 * Added interpolation module to initialisation sequence.
 *
 * Revision 1.6  2001/09/23 18:58:00  brc
 * Added call to support specification of memory sizes in parameter files with
 * suitable semi-official multipliers (k, M, G, T) with their correct binary
 * values (i.e., multiples of 2^10 rather than 1000).
 *
 * Revision 1.5  2001/08/21 00:10:30  brc
 * Added cube() initialisation sequence to call table.
 *
 * Revision 1.4  2001/05/15 01:18:08  brc
 * Modifications to make libccom compile cleanly under Linux.
 *
 * Revision 1.3  2001/05/14 20:46:35  brc
 * Added mapsheet module to call list.
 *
 * Revision 1.2  2001/05/14 04:16:27  brc
 * Added translation routines to map from lengths, durations and rates into
 * uniform MKS units with unit translation and metric multipliers.
 *
 * Revision 1.1  2001/05/13 02:42:15  brc
 * Added facilities to work with a generic parameter file, reading a hierarchically
 * constructed list of modules with string recognition constants for their parameters.
 * The module also has a generic executor so that a command-line user program can
 * have a parameter list passed around all of the modules in the library that
 * are 'params' aware.
 *
 *
 * File:	params.c
 * Purpose:	Parameter reader and parser support code
 * Date:	11 May 2001
 *
 * (c) Center for Coastal and Ocean Mapping, University of New Hampshire, 2001.
 *     This file is unpublished source code of the University of New Hampshire,
 *     and may only be disposed of under the terms of the software license
 *     supplied with the source-code distribution.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <limits.h>
#include <float.h>
#include <math.h>
#include "stdtypes.h"
#include "error.h"
#include "ccom_general.h"
#include "params.h"
#include "att.h"
#include "errmod.h"
#include "nav.h"
#include "stime.h"
#include "tides.h"
#include "vessel.h"
#include "sounding.h"
#include "mapsheet.h"
#include "cube.h"
#include "interp.h"

static char *modname = "params";
static char *modrev = "$Revision: 2 $";

static ParExecutor call_table[] = {
	interp_execute_params,
	att_execute_params,
	errmod_execute_params,
	nav_execute_params,
	stime_execute_params,
	tide_execute_params,
	vessel_execute_params,
	sounding_execute_params,
	mapsheet_execute_params,
	cube_execute_params,
	NULL
};

extern ParList *parser_list;
extern FILE *parin;

/* Routine:	params_search_table
 * Purpose:	Search a ParTable for the given tag
 * Inputs:	*tab	Table to search
 *			*name	Final element to search for in table
 * Outputs:	tag of the element matched in the table, or 0 if not found.
 * Comment:	This is simple linear search, and hence may not be massively
 *			efficient for long parameter lists.
 */

static u32 params_search_table(ParTable *tab, const char *name)
{
	u32		elem;
	
	elem = 0;
	while (tab[elem].name != NULL) {
		if (strcmp(name, tab[elem].name) == 0)
			return(tab[elem].tag);
		++elem;
	}
	return(0);
}

/* Routine:	params_get_tail
 * Purpose:	Extract tail element of a composite string
 * Inputs:	*name	Composite string to work with
 * Outputs:	Returns pointer to last element, or name if no separators are found
 * Comment:	-
 */

static char *params_get_tail(char *name)
{
	char	*rtn;
	
	if ((rtn = strrchr(name, '.')) == NULL) return(name);
	return(rtn+1);
}

/* Routine:	params_match
 * Purpose:	Match the first element in the linked list which has the right
 *			prefix for the calling module
 * Inputs:	*root	Start of linked list to work with
 *			*prefix	Module prefix to match at start of tags
 *			*table	Table of elements to match for this module
 * Outputs:	Returns a pointer to the next element of the linked list to search
 *			from for further matches.
 *			*rtn	Tag of element that was matched
 *			*match	ParList element that was matched
 * Comment:	This provides the main user interface to the parameter matching
 *			algorithm.  This code runs down the list of parameters, finding the
 *			first one which matches the user prefix, and returning a pointer to
 *			the next section of the list to search, the element matched (if any)
 *			and the table element match (if any).  When the list of parameters
 *			is exhausted, the code returns NULL.  Note that the code may return
 *			NULL *and* have a matched tag, since the pointer returned is for
 *			the *next* list element to search.
 */

ParList *params_match(ParList *root, const char *prefix, ParTable *table,
					  u32 *rtn, ParList **match)
{
	u32		prefix_len = strlen(prefix), tag;
	
	*rtn = 0;
	*match = NULL;	/* Failsafe initialisation */
	
	while (root != NULL) {
		if (strncmp(prefix, root->name, prefix_len) == 0) {
			/* Matched prefix, check that remaining component is a leaf */
			if (strchr(root->name+prefix_len+1, '.') == NULL) {
				/* Search supplied table for a match */
				if ((tag = params_search_table(table,
										params_get_tail(root->name))) == 0) {
					/* Prefix matches, but no match in table --- problem */
					error_msgv(modname, "warning: \"%s\" was not matched in the"
						" \"%s\" module parameter list.\n", root->name, prefix);
				} else {
					*rtn = tag;
					*match = root;
					return(root->next);
				}
			}
		}
		root = root->next;
	}
	return(NULL);
}

/* Routine:	params_execute
 * Purpose:	Executive code for all of the parameter modules indicated in the
 *			header: supply parameters to all code.
 * Inputs:	*list	Root of the ParList linked list to use for parameters
 * Outputs:	True if all of the parameter in the list were marked `used' by the
 *			various bits of code; otherwise False.
 * Comment:	The error check makes sure that everything that the user specified
 *			in the parameter file was used.  This also makes sure that if the
 *			user mis-typed something, it gets caught.
 *				This module uses the table at the top of the params.c file to
 *			determine which modules in the library are `params' aware.  That is,
 *			which should be offered the list, and which routine in the module
 *			to call to handle it.  It is up to the callees what they do with the
 *			parameter list, although of course, it would be wierd if they didn't
 *			use params_match() to work through the list.  It is up to the
 *			callee code to pass the list to any sub-modules if required.
 */

Bool params_execute(ParList *list)
{
	u32		elem;
	Bool	rtn = True;
	
	elem = 0;
	while (call_table[elem] != NULL) {
		rtn &= (*call_table[elem])(list);
		++elem;
	}
	while (list != NULL) {
		if (list->used == False) {
			error_msgv(modname, "warning: parameter string \"%s\" was not"
				" claimed by any registered module.\n", list->name);
			rtn = False;
		}
		list = list->next;
	}
	return(rtn);
}

/* Routine:	params_reverse_list
 * Purpose:	Apply the patented ultra-cool Reverse-o-Matic list reverser
 * Inputs:	*start	Head of singly linked list to reverse
 * Outputs:	Returns pointer to new head of linked list
 * Comment:	This reverses the order of the linked list passed using an
 *			algorithm so cunning it could be voted Emeritus Professor of
 *			Cunning at Oxford University.
 */

ParList *params_reverse_list(ParList *start)
{
    ParList *root = NULL, *next;

    while (start != NULL) {
        next = start->next; /* Record where we're going after this */
        start->next = root;
        root = start;       /* Append node to new list */
        start = next;       /* Repoint to next item in previous list */
    }
    return(root);
}

/* Routine:	params_new_from_file
 * Purpose:	Read a parameter list from ASCII file
 * Inputs:	*name	Name of the file to read
 * Outputs:	Returns True if file read OK, otherwise False.
 *			**list	Set to point to linked list of parameter pairs, or NULL
 *					if none read.
 * Comment:	A file with no parameter descriptions is valid (but wierd).
 */

int parparse(void *);

Bool params_new_from_file(const char *name, ParList **list)
{
	if ((parin = fopen(name, "r")) == NULL) {
		error_msgv(modname, "failed to open \"%s\" for parameter list.\n",
			name);
		return(False);
	}
	if (parparse(NULL) < 0) {
		error_msgv(modname, "failed parsing \"%s\" for parameters.\n", name);
		return(False);
	}
	*list = params_reverse_list(parser_list);
	parser_list = NULL;
	return(True);
}

/* Routine:	params_release
 * Purpose:	Release dynamic memory associated with the parameter list
 * Inputs:	*list	List to free
 * Outputs:	-
 * Comment:	-
 */

void params_release(ParList *list)
{
	ParList	*next;
	
	while (list != NULL) {
		next = list->next;
		free(list->name);
		free(list);
		list = next;
	}
}

/* Routine:	params_translate_length
 * Purpose:	Translate a number + units into a usable length measurement
 * Inputs:	*string	String to work with
 * Outputs:	Returns length in meters, or DBL_MAX if conversion fails.
 * Comment:	This code understands a number of different units, which are:
 *				'm'		MKS standard meters
 *				'ft'	Imperial feet (1 ft = 12 in = 0.3048m exactly)
 *				'in'	Imperial inches (1 in = 0.0254m exactly)
 *				'nm'	Nautical miles (1 nm = 1852m)
 *			and where appropriate, MKS scales ('u' = 0.000001, 'm' = 0.001,
 *			'c' = 0.01, 'd' = 0.1, 'h' = 100, 'k' = 1000).  Scales are applied
 *			to the number, converted as a double precision float before the
 *			value is returns, so that the value is always in base unit MKS.
 */

f64 params_translate_length(char *string)
{
	f64		rtn;
	char	*tail, *scale;
	
	if (string == NULL || strlen(string) == 0)
		return(DBL_MAX);
	
	rtn = strtod(string, &tail);
	if (tail == string) {
		error_msgv(modname, "cannot extract a number from \"%s\".\n", string);
		return(DBL_MAX);
	}
	/* Now check whether there is anything left to convert into a unit */
	while (isspace(*tail)) ++tail;
	if (strlen(tail) >= 1) {
		/* Something there --- check known units */
		if (strstr(tail, "nm") != NULL) {
			/* Units in Nautical Miles with no scale conversion (needs to check
			 * for this first to make sure that we don't try to match this as
			 * nano-meters!)
			 */
			rtn *= 1852.0;
		} else if (strstr(tail, "ft") != NULL) {
			/* Units in Imperial feet with no scale conversion */
			rtn *= 0.3048;
		} else if (strstr(tail, "in") != NULL) {
			/* Units in Imperial inches with no scale conversion */
			rtn *= 0.0254;
		} else if ((scale = strrchr(tail, 'm')) != NULL) {
			/* MKS Units with scale conversions */
			switch(*--scale) {
				case 'u':	rtn *= 1e-6; break;
				case 'm':	rtn *= 1e-3; break;
				case 'c':	rtn *= 1e-2; break;
				case 'd':	rtn *= 1e-1; break;
				case 'h':	rtn *= 1e02; break;
				case 'k':	rtn *= 1e03; break;
				default:
					if (!isspace(*scale) && !isdigit(*scale)) {
						error_msgv(modname, "error: unrecognised scale for"
							" length conversion (%c).\n", *scale);
						return(DBL_MAX);
					}
					break;
			}
		} else {
			/* Unknown units? */
			error_msg(modname, "unknown units for length conversion.\n");
			return(DBL_MAX);
		}
	}
	return(rtn);
}

/* Routine:	params_translate_duration
 * Purpose:	Translate a number into units of seconds, with optional scaling
 * Inputs:	*string	String to work from
 * Outputs:	Returns converted string as seconds, or DBL_MAX for failure
 * Comment:	This applies suitable conversions to the numerical part of the
 *			string, and then looks for a units string in the remainder.  The
 *			code recognises common MKS multipliers ('u' = 1e-6, 'm' = '1e-3',
 *			'c' = 1e-2) when the units are seconds ('s'), and also recognises
 *			'min' as minutes, and 'hr' as hours.  Other units of time, and
 *			composites (1 hr 10 min) are not allowed.
 */

f64 params_translate_duration(char *string)
{
	f64		rtn;
	char	*tail, *scale;
	
	if (string == NULL || strlen(string) == 0)
		return(DBL_MAX);
	
	rtn = strtod(string, &tail);
	if (tail == string) {
		error_msgv(modname, "cannot extract a number from \"%s\".\n", string);
		return(DBL_MAX);
	}
	/* Now check whether there is anything left to convert into a unit */
	while (isspace(*tail)) ++tail;
	if (strlen(tail) >= 1) {
		/* Something there --- check known units */
		if (strstr(tail, "hr") != NULL) {
			/* Units in hours */
			rtn *= 60.0 * 60.0;
		} else if (strstr(tail, "mn") != NULL) {
			/* Units in minutes */
			rtn *= 60.0;
		} else if ((scale = strstr(tail, "s")) != NULL) {
			/* MKS Units with scale conversions */
			switch(*--scale) {
				case 'u':	rtn *= 1e-6; break;
				case 'm':	rtn *= 1e-3; break;
				case 'c':	rtn *= 1e-2; break;
				default:
					if (!isspace(*scale) && !isdigit(*scale)) {
						error_msgv(modname, "error: unrecognised scale for"
							" durations (%c).\n", *scale);
						return(DBL_MAX);
					}
					break;
			}
		} else {
			/* Unknown units? */
			error_msg(modname, "unknown units for duration conversion.\n");
			return(DBL_MAX);
		}
	}
	return(rtn);
}

/* Routine:	params_translate_rate
 * Purpose:	Translate a number into units of Hz, including MKS scales.
 * Inputs:	*string	String to work from
 * Outputs:	Returns converted rate in Hz, or DBL_MAX on failure.
 * Comment:	This looks for the units 'Hz', and then applies any scale factors
 *			if knows about ('k' = 1000, 'M' = 1e6) before returning the
 *			converted rate.
 */

f64 params_translate_rate(char *string)
{
	f64		rtn;
	char	*tail, *scale;
	
	if (string == NULL || strlen(string) == 0)
		return(DBL_MAX);
	
	rtn = strtod(string, &tail);
	if (tail == string) {
		error_msgv(modname, "cannot extract a number from \"%s\".\n", string);
		return(DBL_MAX);
	}
	/* Now check whether there is anything left to convert into a unit */
	while (isspace(*tail)) ++tail;
	if (strlen(tail) >= 1) {
		/* Something there --- check known units */
		if ((scale = strstr(tail, "Hz")) != NULL) {
			/* OK - units.  Check for multipliers */
			switch(*--scale) {
				case 'k':
					rtn *= 1000.0;
					break;
				case 'M':
					rtn *= 1.0e6;
					break;
				default:
					if (!isspace(*scale) && !isdigit(*scale)) {
						error_msgv(modname, "error: unrecognised scale for"
							" rate conversion (%c).\n", *scale);
						return(DBL_MAX);
					}
					break;
			}
		} else {
			/* Unknown units? */
			error_msg(modname, "unknown units for duration conversion.\n");
			return(DBL_MAX);
		}
	}
	return(rtn);

}

/* Routine:	params_translate_memory
 * Purpose:	Translate a size in memory units into bytes, including customary
 *			units for powers of two.
 * Inputs:	*string	String to work from
 * Outputs:	Returns converted size in bytes, or INT_MAX on failure
 * Comment:	This looks for the units 'b' == bytes, and then applys any scale
 *			factors it knows about ('k' = 2^10, 'M' = 2^20, 'G' = 2^30, 'T' =
 *			2^40) before returning the size.  Note that since the return value
 *			is only a 32-bit unsigned int, anything over about 4G would overflow
 *			and will be rejected.  The conversion is done internally in floating
 *			point, so fractions of a unit can be specified (e.g., '2.5 Mb'), but
 *			the value is rounded down to an integer number of bytes just before
 *			being returned to the user.
 */

u32 params_translate_memory(char *string)
{
	f64		rtn;
	char	*tail, *scale;
	
	if (string == NULL || strlen(string) == 0)
		return(INT_MAX);
	
	rtn = strtod(string, &tail);
	if (tail == string) {
		error_msgv(modname, "cannot extract a number from \"%s\".\n", string);
		return(INT_MAX);
	}

	/* Now check whether there is anything left to convert into a unit */
	while (isspace(*tail)) ++tail;
	if (strlen(tail) >= 1) {
		/* Something there --- check known units */
		if ((scale = strstr(tail, "b")) != NULL) {
			/* OK - units.  Check for multipliers */
			switch(*--scale) {
				case 'k':
					rtn *= 1024.0;
					break;
				case 'M':
					rtn *= 1024.0*1024.0;
					break;
				case 'G':
					rtn *= 1024.0*1024.0*1024.0;
					break;
				case 'T':
					rtn *= 1024.0*1024.0*1024.0*1024.0;
					break;
				default:
					if (!isspace(*scale) && !isdigit(*scale)) {
						error_msgv(modname, "error: unrecognised scale for"
							" rate conversion (%c).\n", *scale);
						return(INT_MAX);
					}
					break;
			}
		} else {
			/* Unknown units? */
			error_msg(modname, "unknown units for duration conversion.\n");
			return(INT_MAX);
		}
	}
	if (rtn < 0.0 || rtn >= INT_MAX) {
		error_msgv(modname, "error: size is out of representable range"
			"[0, %d).\n", INT_MAX);
		return(INT_MAX);
	}
	return((u32)floor(rtn));

}
